docs/developers.md

```diff
@@ -20,80 +38,297 @@ For completed APB examples, take a look at some of the APBs in the [ansibleplayb
+* `type`: Data type of the parameters as specified by [json-schema](http://json-schema.org/) such as `string`, `number`, `int`, `boolean`, or `enum`.  Default input field type in the UI will be assigned if no `display_type` is assigned.
+* `required`: Whether or not the parameter is required for APB execution.  Required field in UI.
+* `default`: Default value assigned to the parameter.
+* `display_type`: Display type for the UI.  For example, you can override a string input as a `password` to hide it in the UI.  Accepted fields include `text`, `textarea`, `password`, `checkbox`, `select`.
+* `display_group`: will cause a parameter to display in groups with adjacent parameters with matching `display_group` fields.  In the above example, adding another field below with `display_group: Group 1` will visually group them together in the UI under the heading "Group 1".
+
+When using a long list of parameters it might be useful to use a shared parameter list. For an example of this, please see [rhscl-postgresql-apb](https://github.com/ansibleplaybookbundle/rhscl-postgresql-apb/blob/master/apb.yml#L4) for an example.
+
+
+## Dockerfile
+The Dockerfile is what's used to actually build the APB image.  As a result, sometimes you will need to customize it for your own needs.  For example, if running a playbook that requires interactions with PostgreSQL, you may want to install the required packages by adding the `yum install`.
+
+```yaml
+FROM ansibleplaybookbundle/apb-base
+MAINTAINER Ansible Playbook Bundle Community
+
+LABEL "com.redhat.apb.spec"=\
+"<------------base64-encoded-spec------------>"
 
-### Adding optional variables to an Ansible playbook bundle via environment variables
 
-To pass variables into an APB, you will need to escape the variable substitution in your `.yml` files. For example, the below is a section of the [main.yml](https://github.com/fusor/apb-examples/blob/master/etherpad-apb/roles/provision-etherpad-apb/tasks/main.yml#L89) in the [etherpad-apb](https://github.com/fusor/apb-examples/tree/master/etherpad-apb):
+COPY roles /opt/ansible/roles
+COPY playbooks /opt/apb/actions
+RUN chmod -R g=u /opt/{ansible,apb}
 
-```yml
+
+### INSTALL THE REQUIRED PACKAGES
+RUN yum -y install python-boto postgresql && yum clean all
+
+USER apb
+```
+
+## Actions
+An action for an APB is the command that the APB is run with. The 5 standard actions that we support is `provision`, `deprovision`, `bind`, `unbind`, and `test`. For an action to be valid there must be a valid file in the `playbooks` directory named `<action>.yml`. These playbooks can do anything which also means that you can technically create any action you would like. Our [mediawiki-apb](https://github.com/ansibleplaybookbundle/mediawiki123-apb/blob/master/playbooks/update.yml) has an example of creating an action `update`.
+
+Most all APBs will normally have at least a `provision` to create resources and a `deprovision` action to destroy the resources when deleting the service.
```

Comment:
"Most all APBs will..." should read "Most APBs will...". I'd remove the "at least".